--- a/ld/emulparams/elf_i386_mieros.sh	1970-01-01 01:00:00.000000000 +0100
+++ b/ld/emulparams/elf_i386_mieros.sh	2023-01-15 11:40:01.417585271 +0100
@@ -0,0 +1,2 @@
+source_sh ${srcdir}/emulparams/elf_i386.sh
+TEXT_START_ADDR=0x08000000
--- a/ld/emulparams/elf_x86_64_mieros.sh	1970-01-01 01:00:00.000000000 +0100
+++ b/ld/emulparams/elf_x86_64_mieros.sh	2023-01-15 11:40:01.417585271 +0100
@@ -0,0 +1 @@
+source_sh ${srcdir}/emulparams/elf_x86_64.sh
--- a/ld/configure.tgt	2022-07-29 09:37:48.000000000 +0200
+++ b/ld/configure.tgt	2023-01-15 11:40:01.417585271 +0100
@@ -444,6 +444,11 @@
 			;;
 i[3-7]86-*-chaos)	targ_emul=elf_i386_chaos
 			;;
+i[3-7]86-*-mieros*)
+			targ_emul=elf_i386_mieros
+			targ_extra_emuls=elf_i386
+			targ64_extra_emuls="elf_x86_64_mieros elf_x86_64"
+			;;
 ia16-*-elf*)		targ_emul=elf_i386
 			targ_extra_emuls=i386msdos
 			;;
@@ -1053,6 +1058,10 @@
 			targ_extra_emuls=i386pe
 			targ_extra_ofiles="deffilep.o pep-dll.o pe-dll.o"
 			;;
+x86_64-*-mieros*)
+			targ_emul=elf_x86_64_mieros
+			targ_extra_emuls="elf_i386_mieros elf_x86_64 elf_i386"
+			;;
 xgate-*-*)		targ_emul=xgateelf
 			targ_extra_ofiles=ldelfgen.o
 			;;
--- a/ld/Makefile.am	2022-07-08 11:46:48.000000000 +0200
+++ b/ld/Makefile.am	2023-01-15 11:40:01.417585271 +0100
@@ -281,6 +281,7 @@
 	eelf_i386_ldso.c \
 	eelf_i386_sol2.c \
 	eelf_i386_vxworks.c \
+	eelf_i386_mieros.c \
 	eelf_iamcu.c \
 	eelf_s390.c \
 	eh8300elf.c \
@@ -457,6 +458,7 @@
 	eelf_x86_64_fbsd.c \
 	eelf_x86_64_haiku.c \
 	eelf_x86_64_sol2.c \
+	eelf_x86_64_mieros.c \
 	ehppa64linux.c \
 	ei386pep.c \
 	emmo.c
@@ -766,6 +768,7 @@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf32xtensa.Pc@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf32z80.Pc@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_i386.Pc@am__quote@
+@AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_i386_mieros.Pc@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_i386_be.Pc@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_i386_fbsd.Pc@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_i386_haiku.Pc@am__quote@
@@ -941,6 +944,7 @@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf64tilegx_be.Pc@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_mipsel_haiku.Pc@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_x86_64.Pc@am__quote@
+@AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_x86_64_mieros.Pc@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_x86_64_cloudabi.Pc@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_x86_64_fbsd.Pc@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/eelf_x86_64_haiku.Pc@am__quote@
--- a/gas/configure.tgt	2022-07-08 11:46:47.000000000 +0200
+++ b/gas/configure.tgt	2023-01-15 11:40:01.416585290 +0100
@@ -263,6 +263,7 @@
   i386-*-*nt*)				fmt=coff em=pe ;;
   i386-*-rdos*)				fmt=elf ;;
   i386-*-darwin*)			fmt=macho ;;
+  i386-*-mieros*)     fmt=elf em=linux ;;
 
   ia16-*-elf*)				fmt=elf ;;
 
--- a/config.sub	2022-07-08 11:46:47.000000000 +0200
+++ b/config.sub	2023-01-15 11:40:01.416585290 +0100
@@ -1754,7 +1754,7 @@
 	     | onefs* | tirtos* | phoenix* | fuchsia* | redox* | bme* \
 	     | midnightbsd* | amdhsa* | unleashed* | emscripten* | wasi* \
 	     | nsk* | powerunix* | genode* | zvmoe* | qnx* | emx* | zephyr* \
-	     | fiwix* )
+	     | fiwix* | mieros* )
 		;;
 	# This one is extra strict with allowed versions
 	sco3.2v2 | sco3.2v[4-9]* | sco5v6*)
--- a/bfd/config.bfd	2022-07-08 11:46:47.000000000 +0200
+++ b/bfd/config.bfd	2023-01-15 11:40:01.416585290 +0100
@@ -655,7 +655,16 @@
     targ_selvecs=
     targ64_selvecs=x86_64_elf64_vec
     ;;
+  i[3-7]86-*-mieros*)
+    targ_defvec=i386_elf32_vec
+    targ64_selvecs="x86_64_elf64_vec"
+    ;;
 #ifdef BFD64
+  x86_64-*-mieros*)
+    targ_defvec=x86_64_elf64_vec
+    targ_selvecs="i386_elf32_vec"
+    want64=true
+    ;;
   x86_64-*-cloudabi*)
     targ_defvec=x86_64_elf64_cloudabi_vec
     want64=true
